cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStore.java

```diff
@@ -0,0 +1,451 @@
+      // Remove the key from the keystore and throw an IOException.
+      deleteFromStore(keyStore, name, password);
+      throw ioe;
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  private static Key deleteFromStore(KeyStore keyStore, String name, char[] password) throws IOException {
+    Key key = null;
+    try {
+      if (keyStore.containsAlias(name)) {
+        key = keyStore.getKey(name, password);
+        keyStore.deleteEntry(name);
+      }
+    } catch (KeyStoreException | UnrecoverableKeyException | NoSuchAlgorithmException e) {
+      throw new IOException("Failed to delete the key. " + name, e);
+    }
+    return key;
+  }
+
+  /**
+   * Deletes the element with the given name.
+   * @param name Name of the element to be deleted
+   */
+  @Override
+  public void delete(String name) throws IOException {
+    Key key = null;
+    writeLock.lock();
+    try {
+      key = deleteFromStore(keyStore, name, password);
+      // Attempt to persist the store. If this fails then the keystore file will have an extra key
+      // that the in-memory store does not. This will be remedied the next time a flush happens.
+      // If another flush does not happen and the system is restarted, the only time that file is read,
+      // then we will have an extra key.
+      flush();
+    } catch (IOException ioe) {
+      if (key != null) {
+        try {
+          keyStore.setKeyEntry(name, key, password, null);
+        } catch (KeyStoreException e) {
+          throw new IOException("Failed to put back the key after a flush failure.", e);
+        }
+      }
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  /**
+   * List of all the entries in the secure store.
+   * @return A list of {@link SecureStoreMetadata} objects representing the data stored in the store.
+   * @throws IOException
+   */
+  @Override
+  public List<SecureStoreMetadata> list() throws IOException {
+    List<SecureStoreMetadata> list = new ArrayList<>();
+    String name;
+    readLock.lock();
+    try {
+      Enumeration<String> aliases = keyStore.aliases();
+      while (aliases.hasMoreElements()) {
+        name = aliases.nextElement();
+        list.add(getSecureStoreMetadata(name));
+      }
+    } catch (KeyStoreException e) {
+      throw new IOException("Failed to get the list of elements from the secure store.", e);
+    } finally {
+      readLock.unlock();
+    }
+    return list;
+  }
+
+  /**
+   * @param name Name of the data element.
+   * @return An object representing the securely stored data associated with the name.
+   */
+  @Override
+  public SecureStoreData get(String name) throws IOException {
+    readLock.lock();
+    try {
+      if (!keyStore.containsAlias(name)) {
+        throw new IOException(name + " not found in the secure store.");
+      }
+      Key key = keyStore.getKey(name, password);
+      return ((KeyStoreEntry) key).data;
```

Comment:
Better not to access the field directly, even though it's an inner class right now. It makes future refactoring easier.